lib/srv/desktop/rdp/rdpclient/src/client.rs

```diff
@@ -14,129 +14,316 @@ use ironrdp_pdu::rdp::RdpError;
-
-    let upgraded =
-        ironrdp_tokio::mark_as_upgraded(should_upgrade, &mut connector, server_public_key);
-
-    let mut upgraded_framed = ironrdp_tokio::TokioFramed::new(upgraded_stream);
-
-    let connection_result =
-        ironrdp_tokio::connect_finalize(upgraded, &mut upgraded_framed, connector).await?;
-
-    info!("connection_result: {:?}", connection_result);
-
-    let mut x224_processor = Processor::new(
-        connection_result.static_channels,
-        connection_result.user_channel_id,
-        connection_result.io_channel_id,
-        None,
-        None,
-    );
-
-    loop {
-        tokio::select! {
-             res = upgraded_framed.read_pdu() => {
-                let (action, mut frame) = res?;
-                match action {
-                    ironrdp_pdu::Action::X224 => {
-                        let result =
-                            process_x224_frame(&mut x224_processor, &frame).await;
-                        process_active_stage_result(&mut upgraded_framed, result).await.map_err(|e|{
-                                error!("process_stage_result {:?}", e);
-                                ConnectError::Rdp(RdpError::InvalidSecurityHeader)
-                            })?;
-                    },
-                    ironrdp_pdu::Action::FastPath => {
-                            frame_sender.send(frame).await;
-                    },
+    /// Spawns a task for running the RDP loop which:
+    /// 1. Reads new frames from the RDP server and sends them to Go.
+    /// 2. Listens on the Client's function_receiver for function calls
+    ///    which it then executes.
+    ///
+    /// Returns immediately with a receiver which callers are expected to listen
+    /// on in case of any errors, or until a ClientFunction::Stop is received.
+    fn run_rdp_loop(mut self) -> Receiver<ClientError> {
+        let (error_sender, error_receiver) = channel::<ClientError>(1);
+
+        tokio_spawn(async move {
+            macro_rules! handle_error {
+                ($err:expr, $error_sender:expr) => {
+                    $error_sender.send($err.into()).await.unwrap_or_else(|_| {
+                        error!("Failed to send error through the channel");
+                    });
+                    return;
                 };
-             }
-             Some(data) = tdp_reader.recv() => {
-                match data {
-                    TdpMessage::Key(p) => {
-                        write_rdp_key(&mut upgraded_framed, p).await;
+            }
+
+            macro_rules! handle_function {
+                ($self:expr, $func:ident, $args:expr, $error_sender:expr) => {
+                    if let Err(err) = $self.$func($args).await {
+                        handle_error!(err, $error_sender);
+                    }
+                };
+            }
+
+            if let Some(mut function_receiver) = self.function_receiver.take() {
```

Comment:
I'm not a fan of macros here, especially when they do flow control (`return` in `handle_error`). It's hard to read and reason about.
My proposal: extract whole body of this `async` block to separate `fn` returning `Result` and handle sending after return. This way you can use `?` inside and don't need a macro. Check https://github.com/gravitational/teleport/pull/32147, I have similar thing going on

Reply:
Good call

8bc340a1185cc3142040b6a97f29417db50f8200